src/lib/sedlexing.ml

```diff
@@ -80,37 +78,45 @@ let set_position lexbuf position =
 
 let set_filename lexbuf fname = lexbuf.filename <- fname
 
-let fill_buf_from_gen f gen buf pos len =
-  let rec aux i =
-    if i >= len then len
-    else (
-      match gen () with
-        | Some c ->
-            buf.(pos + i) <- f c;
-            aux (i + 1)
-        | None -> i)
-  in
-  aux 0
+let refill_buf_from_gen gen =
+  let malformed = ref false in
+  fun buf pos len ->
+    if !malformed then raise MalFormed;
+    let rec aux i =
+      if i >= len then len
+      else (
+        match gen () with
+          | Some c ->
+              buf.(pos + i) <- c;
+              aux (i + 1)
+          | None -> i
+          | exception MalFormed when i <> 0 ->
+              malformed := true;
+              i)
+    in
+    aux 0
 
-let from_gen s = create (fill_buf_from_gen (fun id -> id) s)
+let from_gen gen = create (refill_buf_from_gen gen)
 
 let from_int_array a =
+  let i = ref 0 in
+  (* copy the array to not be affected by future mutations *)
+  let a = Array.copy a in
```

Comment:
Does our API do mutations? If not, I would instead document that we are not expecting the array to be mutated while being used by the parser and consider those a user error. Copying the whole array will definitely have performance issues with large inputs.